text/0201-refactoring-support.md

```diff
@@ -0,0 +1,193 @@
+# Refactoring Support
+
+* **Original Author(s):**: @rix0rrr
+* **Tracking Issue**: #201
+* **API Bar Raiser**: @RomainMailler
+
+You can now refactor the shape of your construct tree without incurring resource replacement in CloudFormation. To use refactoring support, you must be on `constructs >= 10.x.y`.
+
+## README
+
+When the CDK needs to allocate unique identifiers for your constructs (such as allocating *logical id*s in CloudFormation templates), it will use the full construct path of your construct to base the identifier on. The result of this is that if you refactor any part of your construct tree (for example, if you change a construct identifier, or extract a resource to new construct, thereby introducing a new scope), the generated identifiers will change. This will lead CloudFormation to think a new resource has been introduced and the old one has been removed.
+
+To prevent this from happening, whenever you refactor part of your construct tree you should record the refactoring by calling specific methods indicating what the change was you performed. Let's run through an example.
+
+### Example: extracting a new construct
+
+```ts
+class MyStack extends Stack {
+  constructor(scope: Scope, id: string, props: MyStackProps) {
+    super(scope, id, props);
+
+    new Bucket(this, 'MyBucket');
+  }
+}
+```
+
+Let's say you introduce a new construct called `MyConstruct` to hold the S3 Bucket (and perhaps a couple of other resources). This will make the construct path to the bucket change from `Stack/MyBucket` to `Stack/MyConstruct/MyBucket`, and hence make it allocate a different ID which will lead to CloudFormation trying to delete the Bucket.
+
+When you perform the refactoring, add a call to `this.node.refactor()`:
+
+```ts
+// New extracted construct
+class MyConstruct extends Construct {
+  constructor(scope: Scope, id: string) {
+    super(scope, id);
+
+    new Bucket(this, 'MyBucket');
+  }
+}
+
+// New version of stack construct
+class MyStack extends Stack {
+  constructor(scope: Scope, id: string, props: MyStackProps) {
+    super(scope, id, props);
+
+    new MyConstruct(this, 'MyConstruct');
+    this.node.refactor(this, 'MyBucket', 'MyConstruct/MyBucket');
```

Comment:
API thoughts:

* The term "refactor" doesn't necessarily mean "rename". Perhaps be a bit more explicit about the name of this method. 
* It might be useful to *also* support directly referencing the target object (less potential for mistakes based on string matching)
* Why do we need `this` as the first argument if we are already in the `node`?

```ts
this.node.moveToPath("MyBucket", "MyConstruct/MyBucket");
this.node.moveTo("MyBucket", myConstruct.myBucket);
```